### PR TITLE
[bitnami/redmine] Add Redmine Helm Chart Test

### PR DESCRIPTION
**Pull request by superaleks**
Signed-off-by: alukic <alukic@vmware.com>

**Description of the change**
Increasing the test coverage of the Redmie Helm Chart by adding Cypress and Goss tests. 

**Benefits**

Ensuring higher quality of the Helm charts

**Possible drawbacks**
Automated tests could show to be potentially flaky. 

**Test results** - IN PROGRESS

9 consecutive pipelines test is in progress
[Please see them in the history here](https://github.com/superaleks/charts/pull/16)

**Checklist**

- [X] Title of the PR starts with chart name (e.g. [bitnami/<name_of_the_chart>])
- [X] All commits signed off and in agreement of [Developer Certificate of Origin (DCO)](https://github.com/bitnami/charts/blob/master/CONTRIBUTING.md#sign-your-work)

### CODE REVIEW

File: .vib/redmine/cypress/cypress/plugins/index.js [hunk 1]
```diff
@@ -0,0 +1,22 @@
+/// <reference types="cypress" />
```
Comment (FraPazGal) [inline .vib/redmine/cypress/cypress/plugins/index.js:1]: This file can be removed.

File: .vib/redmine/cypress/cypress/fixtures/issues.json [hunk 2]
```diff
@@ -0,0 +1,7 @@
+{
+  "newIssue": {
+    "tracker": "Bug",
```
Comment (FraPazGal) [inline .vib/redmine/cypress/cypress/fixtures/issues.json:3]: It seems we are not using `newIssue.tracker`.

File: .vib/redmine/cypress/cypress/integration/redmine_spec.js [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+  cy.contains('Issues').click();
+  cy.get('.icon-add').click();
+  cy.fixture('issues').then((issue) => {
+    cy.get('#issue_subject').type(`${issue.newIssue.subject}.${random}`);
+    cy.get('#issue_description').type(issue.newIssue.description);
+    cy.get('input[type="file"]').selectFile('cypress/fixtures/issues.json');
+  });
+  cy.get('[name="commit"]').click();
+  cy.contains('#flash_notice', 'created');
+  cy.contains('issues.json');
```
Comment (FraPazGal) [inline .vib/redmine/cypress/cypress/integration/redmine_spec.js:48]: Would this `contains` succeed if the previous step failed? My concern is that if it stays on the previous page, it will detect the string `issues.json` from the uploaded file to be committed.

File: .vib/redmine/cypress/cypress/fixtures/roles.json [hunk 4]
```diff
@@ -0,0 +1,5 @@
+{
```
Comment (FraPazGal) [inline .vib/redmine/cypress/cypress/fixtures/roles.json:1]: I may be missing something, are we using this fixture somewhere?

File: .vib/redmine/cypress/cypress/integration/redmine_spec.js [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+  });
+  cy.contains('[type="submit"]', 'Save').click({ force: true });
+  cy.contains('Successful update.');
+});
+
+it('allows admin to verify and change application configuration', () => {
+  cy.login();
+  cy.visit('/admin');
+  cy.contains('Settings').click();
+  cy.get('#settings_app_title').clear().type(`VMWare.${random}`);
```
Comment (FraPazGal) [inline .vib/redmine/cypress/cypress/integration/redmine_spec.js:74]: Should we obtain this value from a `settings.json` fixture?

File: .vib/redmine/cypress/cypress/integration/redmine_spec.js [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+  cy.contains('input', 'Save').click();
+  cy.contains('#flash_notice', 'Successful update.');
+  cy.visit('/admin/info');
+  cy.contains('Mysql');
+  cy.contains('production');
+  cy.contains('Git');
```
Comment (FraPazGal) [inline .vib/redmine/cypress/cypress/integration/redmine_spec.js:80]: I'm not familiar with the asset but shouldn't we use `const` variables for these values?

**Review by FraPazGal — CHANGES_REQUESTED** [review]
Thanks for the PR @superaleks, could you please check my comments?

File: .vib/redmine/cypress/cypress/fixtures/issues.json (hunk 2 above)
Comment (superaleks) [reply .vib/redmine/cypress/cypress/fixtures/issues.json]: True. I added the test data thinking I'll need it. Then I forgot to remove it. Now it's fixed.

File: .vib/redmine/cypress/cypress/integration/redmine_spec.js (hunk 3 above)
Comment (superaleks) [reply .vib/redmine/cypress/cypress/integration/redmine_spec.js:48]: I've tried this and this is my conclusion
We have two assertions here:
```
cy.contains('#flash_notice', 'created');
cy.contains('issues.json');
```
If the file is not uploaded properly assertion 2 will fail since it will not find the `issues.json` on the page. 
If the bug is not created, assertion 1 will fail since the bug will not be created. 
So, if the bug is not committed we will not arrive at the file upload test since the first assertion will fail beforehand. This, to me, looks like expected behaviour and doesn't allow for the corner case you worry about.

File: .vib/redmine/cypress/cypress/integration/redmine_spec.js (hunk 5 above)
Comment (superaleks) [reply .vib/redmine/cypress/cypress/integration/redmine_spec.js]: True. I've moved this to the fixture.

File: .vib/redmine/cypress/cypress/integration/redmine_spec.js (hunk 6 above)
Comment (superaleks) [reply .vib/redmine/cypress/cypress/integration/redmine_spec.js]: After your comment, I was indecisive about using `const` or fixtures.
Since we already have the `settings.json` file, I think it makes more sense to have this there. I've applied this change as well.

Comment (FraPazGal) [reply .vib/redmine/cypress/cypress/integration/redmine_spec.js]: Maybe a bit unrelated to this, is it necessary to check the page contains these 3 elements? I could be missing something as I'm not familiar with redmine, but we already checked the "Successful update" notice and these elements look like were already as they are before the update.

File: .vib/redmine/cypress/cypress/fixtures/smtps.json [hunk 7]
```diff
@@ -0,0 +1,7 @@
+{
+  "newSMTP": {
+    "smtpEmailAddress": "redmine@vmware.net",
```
Comment (joancafom) [inline .vib/redmine/cypress/cypress/fixtures/smtps.json:3]: I'd rather use a truly fake one, just in case. e.g. `redmine@example.com`

File: .vib/redmine/cypress/cypress/fixtures/users.json [hunk 8]
```diff
@@ -0,0 +1,9 @@
+{
+  "newUser": {
+    "userName": "JaneJessica",
+    "password": "someComplicatedPass12345!",
+    "firstName": "Jane Doe Jess",
+    "lastName": "Doe",
+    "email": "thisisanemail"
```
Comment (joancafom) [inline .vib/redmine/cypress/cypress/fixtures/users.json:7]: Is this a valid email?

File: .vib/redmine/cypress/cypress/integration/redmine_spec.js [hunk 9]
```diff
@@ -0,0 +1,86 @@
+/// <reference types="cypress" />
+import { random } from './utils';
+
+it('allows admin to login/logout', () => {
+  cy.login();
+  cy.get('#flash_error').should('not.exist');
```
Comment (joancafom) [inline .vib/redmine/cypress/cypress/integration/redmine_spec.js:6]: When possible, rather check the error message than the element's id, e.g.:

```
cy.contains('Invalid user').should('not.exist');
```

File: .vib/redmine/cypress/cypress/integration/redmine_spec.js [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+  cy.visit('account/register');
+  cy.fixture('users').then((user) => {
+    cy.get('#user_login').type(`${user.newUser.userName}.${random}`);
+    cy.get('#user_password').type(`${user.newUser.password}.${random}`);
+    cy.get('#user_password_confirmation').type(
+      `${user.newUser.password}.${random}`
+    );
+    cy.get('#user_firstname').type(`${user.newUser.firstName}.${random}`);
+    cy.get('#user_lastname').type(`${user.newUser.lastName}.${random}`);
+    cy.get('#user_mail').type(`${user.newUser.email}.${random}@email.com`);
```
Comment (joancafom) [inline .vib/redmine/cypress/cypress/integration/redmine_spec.js:21]: Now I understand why the domain was not in the fixtures. Isn't it easier something like `${random}${user.newUser.email}`, though?

File: .vib/redmine/cypress/cypress/integration/redmine_spec.js [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+
+it('allows admin to create a project and an issue with file uploaded', () => {
+  cy.login();
+  cy.visit('projects');
+  cy.contains('New project').click();
+  cy.fixture('projects').then((project) => {
+    cy.get('#project_name').type(`${project.newProject.name}.${random}`);
+    cy.get('#project_description').type(project.newProject.description);
+    cy.get('[name="commit"]').click();
+    cy.contains('Successful creation.');
```
Comment (joancafom) [inline .vib/redmine/cypress/cypress/integration/redmine_spec.js:35]:
```suggestion
    cy.contains('Successful creation');
```

File: .vib/redmine/cypress/cypress/integration/redmine_spec.js [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+
+it('allows admin to login/logout', () => {
+  cy.login();
+  cy.get('#flash_error').should('not.exist');
+  cy.contains('Sign out').click();
+  cy.contains('Sign in');
+});
+
+it('allows user to register', () => {
+  cy.visit('account/register');
```
Comment (joancafom) [inline .vib/redmine/cypress/cypress/integration/redmine_spec.js:12]: You're interchangeably using `cy.visit` with a leading `/` and without it. Use only one approach:

```suggestion
  cy.visit('/account/register');
```

File: .vib/redmine/cypress/cypress/integration/redmine_spec.js [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+    cy.get('#user_lastname').type(`${user.newUser.lastName}.${random}`);
+    cy.get('#user_mail').type(`${user.newUser.email}.${random}@email.com`);
+  });
+  cy.get('[type="submit"]').click();
+  cy.contains('Your account was created');
+});
+
+it('allows admin to create a project and an issue with file uploaded', () => {
+  cy.login();
+  cy.visit('projects');
```
Comment (joancafom) [inline .vib/redmine/cypress/cypress/integration/redmine_spec.js:29]:
```suggestion
  cy.visit('/projects');
```

File: .vib/redmine/cypress/cypress/integration/redmine_spec.js [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+    cy.get('#issue_subject').type(`${issue.newIssue.subject}.${random}`);
+    cy.get('#issue_description').type(issue.newIssue.description);
+    cy.get('input[type="file"]').selectFile('cypress/fixtures/issues.json');
+  });
+  cy.get('[name="commit"]').click();
+  cy.contains('#flash_notice', 'created');
+  cy.contains('issues.json');
+});
+
+it('allows admin to check smtp configuration', () => {
```
Comment (joancafom) [inline .vib/redmine/cypress/cypress/integration/redmine_spec.js:51]: This is not 'checking' but rather 'setting', right?

File: .vib/redmine/cypress/cypress/integration/redmine_spec.js [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+    cy.get('#settings_emails_header')
+      .scrollIntoView()
+      .clear()
+      .type(`${smtp.newSMTP.header}.${random}`);
+    cy.get('#settings_emails_footer')
+      .clear()
+      .type(`${smtp.newSMTP.footer}.${random}`);
+  });
+  cy.contains('[type="submit"]', 'Save').click({ force: true });
+  cy.contains('Successful update.');
```
Comment (joancafom) [inline .vib/redmine/cypress/cypress/integration/redmine_spec.js:67]:
```suggestion
  cy.contains('Successful update');
```

File: .vib/redmine/cypress/cypress/integration/redmine_spec.js [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+  cy.login();
+  cy.visit('/admin');
+  cy.contains('Settings').click();
+  cy.fixture('settings').then((setting) => {
+    cy.get('#settings_app_title')
+      .clear()
+      .type(`${setting.newSetting.appTitle}.${random}`);
+
+    cy.contains('input', 'Save').click();
+    cy.contains('#flash_notice', 'Successful update.');
```
Comment (joancafom) [inline .vib/redmine/cypress/cypress/integration/redmine_spec.js:80]:
```suggestion
    cy.contains('#flash_notice', 'Successful update');
```

File: .vib/redmine/cypress/cypress/integration/redmine_spec.js [hunk 17]
```diff
@@ -0,0 +1,86 @@
...
+    cy.visit('/admin/info');
+    cy.contains(setting.newSetting.database);
+    cy.contains(setting.newSetting.environment);
+    cy.contains(setting.newSetting.versionControl);
```
Comment (joancafom) [inline .vib/redmine/cypress/cypress/integration/redmine_spec.js:84]: You are not updating those fields in the previous step, how can the new settings be applied then?

File: .vib/redmine/cypress/cypress/integration/redmine_spec.js (hunk 6 above)
Comment (superaleks) [reply .vib/redmine/cypress/cypress/integration/redmine_spec.js]: The intention of the test is to **verify**(observe) the application configuration, as well as modify it. 
When we modify the configuration, we go to the info page which is under the admin tab. 
The tests are taken a look at this page:

redmine

Since they are showing important components of this application (database, version control) I thought it is relevant to have them.

File: .vib/redmine/cypress/cypress/fixtures/smtps.json (hunk 7 above)
Comment (superaleks) [reply .vib/redmine/cypress/cypress/fixtures/smtps.json:3]: Why, what is the difference?

File: .vib/redmine/cypress/cypress/integration/redmine_spec.js (hunk 14 above)
Comment (superaleks) [reply .vib/redmine/cypress/cypress/integration/redmine_spec.js:51]: Well, it's doing both.
The SMTP input form doesn't appear if the SMTP is not enabled in the configuration (runtime parameters).
The admin can change the email settings, but cannot set them.
I think this will be clearer if I rename the tests to
`allows admins to verify SMTP is enabled and modify it`

File: .vib/redmine/cypress/cypress/integration/redmine_spec.js (hunk 9 above)
Comment (superaleks) [reply .vib/redmine/cypress/cypress/integration/redmine_spec.js:6]: In the case of error messages within consistent UI error elements, I disagree. Especially for login.
Imagine that the login test fails for something **other** than username and password being incorrect.
The error message will be different. 
But, the UI element (flash error) showing the error message will still be the same since it's used consistently in the application.
If we limit ourselves to discovering error messages only, we will catch only one type of issue.
WDTY? 

Screenshot 2022-05-11 at 11 57 59

File: .vib/redmine/cypress/cypress/integration/redmine_spec.js (hunk 17 above)
Comment (superaleks) [reply .vib/redmine/cypress/cypress/integration/redmine_spec.js:84]: The test is named like this

`it('allows admin to modify and observe application configuration'`

This is because I am not only modifying the application configuration but also verifying the existing one. [Please check a more elaborate explanation I gave to Fran in a separate thread. ](https://github.com/bitnami/charts/pull/10039#discussion_r870100005)

File: .vib/redmine/cypress/cypress/integration/redmine_spec.js (hunk 10 above)
Comment (superaleks) [reply .vib/redmine/cypress/cypress/integration/redmine_spec.js]: Yeah, fair enough! I fixed it, which means both of the threads related to emails are resolved.

File: .vib/redmine/cypress/cypress/fixtures/users.json (hunk 8 above)
Comment (superaleks) [reply .vib/redmine/cypress/cypress/fixtures/users.json]: Made the change requested in a separate comment.

File: .vib/redmine/cypress/cypress/fixtures/smtps.json (hunk 7 above)
Comment (joancafom) [reply .vib/redmine/cypress/cypress/fixtures/smtps.json:3]: I am usually not very into using real domains, just in case haha but feel free to leave it as it is. It was just a personal preference

File: .vib/redmine/cypress/cypress/integration/redmine_spec.js (hunk 9 above)
Comment (joancafom) [reply .vib/redmine/cypress/cypress/integration/redmine_spec.js]: I agree with you here! My concern was that text is less prone to change, but I totally see your point

File: .vib/redmine/cypress/cypress/integration/redmine_spec.js (hunk 17 above)
Comment (joancafom) [reply .vib/redmine/cypress/cypress/integration/redmine_spec.js:84]: Ok, now I see it. However, when you first read this it gives the impression that those are `newSetting`s to be updated in the application. I'd move them to another fixture, say `defaultSetting`.

File: .vib/redmine/goss/goss.yaml [hunk 18]
```diff
@@ -0,0 +1,40 @@
...
+    group: root
+    contains:
+      - "postgresql"
+      - "redmine"
+  /bitnami/redmine/files:
+    mode: "0775"
+    filetype: directory
+    exists: true
+    group: root
+  /opt/bitnami/mysql:
```
Comment (joancafom) [inline .vib/redmine/goss/goss.yaml:31]: Not a big deal, but there are 6 tests here

**Review by joancafom — APPROVED** [review]
LGTM
